### PR TITLE
CLDR-14588 fix known issues output, reduce test noise

### PR DESCRIPTION
**Pull request by srl295**
CLDR-14588 fix known issues output, reduce test noise

- known issues (ICU and CLDR) are now valid urls once more
- quiet abstractcache errors (don't need full stacktrace for missing file)
- quiet a number of other tests, logger available from getLogger()
- only show n (default: 2) lines of 'known issue' output, to keep the output manageable
All output is still available using for example:
 `mvn --file=tools/pom.xml test -pl cldr-code '-Dorg.unicode.cldr.unittest.testArgs=-allKnownIssues -n -q'`

- [ ] This PR completes the ticket.

With this fix, output is as below. Note that excessive known issues are rolled up, but `-allKnownIssues` will show all lines.

```
 13 Known Issues:
CLDR-11583 <https://unicode-org.atlassian.net/browse/CLDR-11583>
  - CLDR/TestBasic/TestBasicDTDCompatibility (Comment out test until last release data is available for unit tests)
  - CLDR/TestBasic/TestDtdCompatibility (Comment out test until last release data is available for unit tests)
CLDR-13313 <https://unicode-org.atlassian.net/browse/CLDR-13313>
  - CLDR/TestTransforms/TestCasing (Investigate the Lithuanian casing test, it may be wrong)
CLDR-13660 <https://unicode-org.atlassian.net/browse/CLDR-13660>
  - CLDR/TestAnnotations/testEmojiOrdering (slightly out of order)
CLDR-13920 <https://unicode-org.atlassian.net/browse/CLDR-13920>
  - CLDR/TestUnits/TestDerivedCase (finish this as part of unit derivation work)
CLDR-13951 <https://unicode-org.atlassian.net/browse/CLDR-13951>
  - CLDR/TestCoverageLevel/testLogicalGroupingSamples (Add more LogicalGrouping tests, fix DECIMAL_FORMAT_LENGTH, etc.)
CLDR-14166 <https://unicode-org.atlassian.net/browse/CLDR-14166>
  - CLDR/LanguageInfoTest/TestChinese (Skip until CLDR updated for new ICU4J LocaleMatcher)
  ... and 26 more
CLDR-6342 <https://unicode-org.atlassian.net/browse/CLDR-6342>
  - CLDR/TestExampleGenerator/TestAllPaths (Need an example for each path used in context: //ldml/localeDisplayNames/subdivisions/subdivision[@type="([^"]*+)"])
  ... and 68 more
CLDR-7075 <https://unicode-org.atlassian.net/browse/CLDR-7075>
  - CLDR/TestSupplementalInfo/TestPluralCompleteness (Missing ordinal minimal pairs)
  - CLDR/TestSupplementalInfo/TestPluralSamples2 (Missing ordinal minimal pairs)
CLDR-7839 <https://unicode-org.atlassian.net/browse/CLDR-7839>
  - CLDR/TestSupplementalInfo/TestPluralCompleteness (Missing plural data for modern locales)
  - CLDR/TestSupplementalInfo/TestPluralRanges (Missing plural data for modern locales)
CLDR-8909 <https://unicode-org.atlassian.net/browse/CLDR-8909>
  - CLDR/TestDtdData/TestValueAttributesWithChildren (waiting for RBNF to use data)
CLDR-9784 <https://unicode-org.atlassian.net/browse/CLDR-9784>
  - CLDR/TestPaths/TestNonLdml (fix TODO's in Attribute validity tests)
CLDR-9982 <https://unicode-org.atlassian.net/browse/CLDR-9982>
  - CLDR/TestDtdData/TestValueAttributesWithChildren (Lower priority fixes to bad xml)
ICU-21241 <https://unicode-org.atlassian.net/browse/ICU-21241>
  - CLDR/LanguageInfoTest/testFallbacks (waiting on LocaleMatcherData update)
  ... and 6 more
 (Use -allKnownIssues to show all known issue sites) 
```

### CODE REVIEW

**Comment by markusicu** [comment]
> With this fix, output is as below.

Why does it still show Trac ticket links?

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java [hunk 1]
```diff
@@ -27,13 +27,18 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.logging.Logger;
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:30]: Are you planning to sync the changes here back into https://github.com/unicode-org/icu/blob/main/icu4j/main/tests/framework/src/com/ibm/icu/dev/test/TestFmwk.java ?

That would be good...

And what about the equivalent feature in ICU4C tests?

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java [hunk 2]
```diff
@@ -846,18 +890,28 @@ public void msg(String message, int level, boolean incCount, boolean newln) {
...
     }
 
-    static final String ICU_TRAC_URL = "http://bugs.icu-project.org/trac/ticket/";
-    static final String CLDR_TRAC_URL = "http://unicode.org/cldr/trac/ticket/";
-    static final String CLDR_TICKET_PREFIX = "cldrbug:";
+    static final Pattern ICU_TICKET_PATTERN = PatternCache.get(
+        "(?i)(?:icu-)?(\\d+)"
+    );
+    static final Pattern CLDR_TICKET_PATTERN = PatternCache.get(
+        "(?i)cldr(?:bug:)?(?:-)(\\d+)"
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:897]: Doesn't this require a dash even after `cldrbug:` ?

**Review by markusicu — COMMENTED** [review]
PR title & commit message sound like you are actually *fixing known issues* -- but you are “only” fixing their output -- maybe rephrase those?

**Comment by srl295** [comment]
> > With this fix, output is as below.
> 
> Why does it still show Trac ticket links?

I will include the latest output

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java (hunk 1 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:30]: Is it still used in ICU4J? I thought those were ported over.

**Comment by srl295** [comment]
Updated output:

```
 6 Known Issues:
CLDR-13660 <https://unicode-org.atlassian.net/browse/CLDR-13660>
  - CLDR/TestAnnotations/testEmojiOrdering (slightly out of order)
CLDR-13920 <https://unicode-org.atlassian.net/browse/CLDR-13920>
  - CLDR/TestUnits/TestDerivedCase (finish this as part of unit derivation work)
CLDR-13951 <https://unicode-org.atlassian.net/browse/CLDR-13951>
  - CLDR/TestCoverageLevel/testLogicalGroupingSamples (Add more LogicalGrouping tests, fix DECIMAL_FORMAT_LENGTH, etc.)
CLDR-14166 <https://unicode-org.atlassian.net/browse/CLDR-14166>
  - CLDR/LanguageInfoTest/TestChinese (Skip until CLDR updated for new ICU4J LocaleMatcher)
  ... and 26 more
ICU-21241 <https://unicode-org.atlassian.net/browse/ICU-21241>
  - CLDR/LanguageInfoTest/testFallbacks (waiting on LocaleMatcherData update)
  ... and 6 more
Unknown Ticket <https://unicode-org.atlassian.net/browse/Unknown Ticket>
  - CLDR/TestBasic/TestBasicDTDCompatibility (Comment out test until last release data is available for unit tests)
  ... and 78 more
 (Use -allKnownIssues to show all known issue sites) 
```

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java (hunk 1 above)
Comment (markusicu) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:30]: > Is it still used in ICU4J? I thought those were ported over to junit.

Yes, TestFmwk.java is still there, and is used. The ICU version *calls* JUnit.

**Comment by srl295** [comment]
@markusicu updated the commit message also

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java (hunk 1 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:30]: hm. TestFmwk.java in ICU is _quite_ divergent.  CLDR picked up 57.1, probably pre JUnit.

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java (hunk 2 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java]: fixed

**Comment by srl295** [comment]
refactored for inclusion into ICU4J (re-integrating TestFmwk completely is an exercise for the reader… )

**Comment by macchiati** [comment]
It didn't have high enough ROI for us to spend the time on. I think it took
Stewart about 6 months to fully convert the ICU test and we have a lot more
important things to do with 6 months!

Mark (📱)

**Comment by markusicu** [comment]
> Updated output:

Could you please update the top PR description with this?

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java [hunk 3]
```diff
@@ -1395,6 +1383,8 @@ public void init() {
             invalidCount = 0;
             testCount = 0;
             random = seed == 0 ? null : new Random(seed);
+
+            knownIssues = new UnicodeKnownIssues(allKnownIssues);
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:1387]: optional: maybe skip this if -prop:logKnownIssue=no ?

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java [hunk 4]
```diff
@@ -1206,6 +1191,7 @@ public void write(String s, int off, int len) {
         public Random random;
         public int maxTargetSec = 10;
         public HashMap props;
+        private UnicodeKnownIssues knownIssues;
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:1194]: nice!

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    /**
+     * Base URL for browsing Unicode JIRA
+     */
+    public static final String UNICODE_JIRA_BROWSE = "https://unicode-org.atlassian.net/browse/";
+
+    static final Pattern ICU_TICKET_PATTERN = Pattern.compile(
+        "(?i)(?:icu-)?(\\d+)"
+    );
+    static final Pattern CLDR_TICKET_PATTERN = Pattern.compile(
+        "(?i)cldr(?:bug:)?(?:-)?(\\d+)"
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:49]:
```suggestion
        "(?i)cldr(?:bug:|-)?(\\d+)"
```

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+     * known issue list
+     * @param ticket A ticket number string. For an ICU ticket, use "ICU-10245".
+     * For a CLDR ticket, use "CLDR-12345".
+     * For compatibility, "1234" -> ICU-1234 and "cldrbug:456" -> CLDR-456
+     * @param comment Additional comment, or null
+     *
+     */
+    public void logKnownIssue(String path, String ticket, String comment) {
+
+        StringBuffer descBuf = new StringBuffer(path);
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:74]:
```suggestion
        StringBuilder descBuf = new StringBuilder(path);
```

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+        if (comment != null && comment.length() > 0) {
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:76]:
```suggestion
        if (!Strings.isNullOrEmpty(comment)) {
```

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            descBuf.append(" (" + comment + ")");
+        }
+        String description = descBuf.toString();
+
+        String ticketLink = "Unknown Ticket";
+        if (ticket != null && ticket.length() > 0) {
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:82]:
```suggestion
        if (!Strings.isNullOrEmpty(ticket)) {
```
https://guava.dev/releases/19.0/api/docs/com/google/common/base/Strings.html#isNullOrEmpty(java.lang.String)

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java (hunk 3 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/TestFmwk.java:1387]: could, but:
- getBooleanProperty is actually not callable from this line as it's in the owning class
-  it's one object per process which contains a map
- logKnownIssue=yes is the default

File: tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java (hunk 8 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:81]: is that available for ICU also?

Comment (srl295) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:81]: I was trying to keep the dependencies small here, seeing as how it already needs a hack to support java 1.7

Comment (srl295) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:81]: i don't see any other use of Strings besides one call to Strings.repeat(), is this a commonly used function?

Comment (markusicu) [reply tools/cldr-code/src/main/java/com/ibm/icu/dev/test/UnicodeKnownIssues.java:81]: These things are common in Google code :-)
But good call on keeping it easy to sync into ICU where we are still on Java 7 (working on it...) and don't use Guava.

**Review by markusicu — APPROVED** [review]
lgtm tnx
